File: src/com/leetcode/array/M15_Three_Sum.java
package com.leetcode.array;

import java.util.*;

/**
 * Created by Michael on 2016/10/24.
 * Given an array S of n integers, are there elements a, b, c in S such that a + b + c = 0?
 * Find all unique triplets in the array which gives the sum of zero.
 * The solution set must not contain duplicate triplets.
 *
 * For example, given array S = [-1, 0, 1, 2, -1, -4],
 * A solution set is:
 * [
 *  [-1, 0, 1],
 *  [-1, -1, 2]
 * ]
 *
 * Function Signature:
 * public List<List<Integer>> threeSum(int[] a) {...}
 *
 * <K-Sum系列问题>
 *    E1 2-Sum: 给定一个整型数组a和目标值k，求解相加等于k的两个元素的索引。（有且仅有一个解）
 *   M15 3-Sum: 给定一个整形数组a和目标值0，求解所有相加等于0的三元组，不可重复。（解个数随机）
 *   M18 4-Sum: 给定一个整型数组a和目标值k，求解所有相加等于0的四元组，不可重复。（解个数随机）
 *  M167 2-Sum Sorted:
 *  E170 2-Sum Data Structure:
 *
 * <Tags>
 * - Sort + Two Pointers: [left → → → ... ← ← ← right]
 *
 */
public class M15_Three_Sum {
    public static void main(String[] args) {
        List<List<Integer>> result = threeSum2(new int[] {-1, 0, 1, 2, -1, -4});    // Should return 2 results.
        List<List<Integer>> result2 = threeSum2(new int[] {0, 0, 0, 0});   // Should return 1 result.
        System.out.println(result.size());
        System.out.println(result2.size());
    }

    /** 最佳解法：Sort + Two Pointers
     *  标准的 “先排序后解决问题”。因为常规解法需要o(n^4)，已经远远高过排序本身的复杂度o(nlogn)
     *  因此有必要先排序，再使用排序所带了的很好的性质，使得最后整体复杂度降为了o(n^2)。爽爽的。
     *  将2Sum的解法扩展到这里。外循环遍历第一个值，扫描范围是0到倒数第三个元素（给后面两个元素留位置）
     *  内循环的target就是(0 - 第一个值)
     *  内循环用双指针扫描，从而实现在o(n)复杂度下达到既不丢解，也不重复记录解。
     */
    // 已排序数组的特性1：任意两个元素之和依旧已排序。可以根据这个左右移动首尾指针，记录所有独特解。
    // 已排序数组的特性2：任何相等的元素都一定会相邻，因此可以直接跳过。根据这个性质避免记录重复解。
    // 为了进一步避免重复解，在从左向右遍历三元组第一成员的时候，应该限制第二和第三成员的搜索范围在第一成员的<右侧>。
    // 简化代码：使用Arrays.asList()直接生成一个填满元素的List对象。而不用使用add()语句好几次。
    static List<List<Integer>> threeSum2(int[] a) {
        List<List<Integer>> result = new ArrayList<>();
        Arrays.sort(a);
        for (int i = 0; i < a.length - 2; i++) {            // 循环至右侧不够两个元素
            if (i > 0 && a[i] == a[i - 1]) continue;        // 跳过重复元素：三元组第一个成员
            int left = i + 1;
            int right = a.length - 1;
            int target = 0 - a[i];
            while (left < right) {
                if      (a[left] + a[right] > target) right--;
                else if (a[left] + a[right] < target) left++;
                else {
                    result.add(new ArrayList<>(Arrays.asList(a[i], a[left], a[right])));
                    left++; right--;                                            // 内循环用的是while，因此必须手动移动两个指针，否则会死循环
                    while (left < right && a[left] == a[left - 1]) left++;      // 跳过重复元素：三元组第二个成员
                    while (left < right && a[right] == a[right + 1]) right--;   // 跳过重复元素：三元组第三个成员
                }
            }
        }
        return result;
    }

    // 下面是错误解法：试图将2Sum的HashMap解法用在这里，
    // 但是HashMap解法只能在仅有唯一解的前提下才能正常工作，因此无法避免重复。
    static List<List<Integer>> threeSumWrong(int[] a) {
        List<List<Integer>> result = new ArrayList<>();
        if (a.length < 3) return result;
        Arrays.sort(a);
        for (int i = 0; i < a.length - 2; i++) {
            if (i > 0 && a[i] == a[i - 1]) continue;        // Avoid Duplicate
            int target = -a[i];
            Map<Integer, Integer> map = new HashMap<>();
            for (int j = i + 1; j < a.length; j++) {
                for (int k = j + 1; k < a.length; k++) {
                    if (map.containsKey(a[j])) {
                        List<Integer> set = new ArrayList<>();
                        set.add(a[i]);
                        set.add(target - a[j]);
                        set.add(a[j]);
                        result.add(set);
                        break;              // 错误：无法做到既确保每个不同解都记录下来，又确保不重复记录相同解。
                    } else map.put(target - a[j], a[j]);
                }
            }
        }
        return result;
    }

    // 2Sum解法供参考。主要区别：
    // 1. 2Sum是直接给你提供target，而3Sum是你在遍历的时候主动选择的值。
    // 2. 2Sum需要返回的是索引值，而3Sum需要返回的是元素值本身。
    // 3. 2Sum告诉你只会有一个解，不需要处理多解情况，而3Sum则需要处理多解但并不重复的情况。
    static int[] twoSum(int[] a, int target) {
        Map<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < a.length; i++) {
            if (map.containsKey(a[i])) return new int[]{map.get(a[i]), i};
            else map.put(target - a[i], i);
        }
        return new int[] {};
    }


    /** 穷举法, time - o(n^4) */
    // 遍历三个元素就用了三个for循环，
    // 同时还需要检查待插入结果是否已经重复，这又是一个for循环，所以整体时间复杂度是o(n^4)
    static List<List<Integer>> threeSum(int[] a) {
        List<List<Integer>> result = new ArrayList<>();
        Set<Integer> pattern = new HashSet<>();
        if (a.length < 3) return result;
        for (int x = 0; x < a.length - 2; x++) {
            for (int y = x + 1; y < a.length - 1; y++) {
                for (int z = y + 1; z < a.length; z++) {
                    if (a[x] + a[y] + a[z] == 0) {
                        List<Integer> set = new ArrayList<>();
                        set.add(a[x]);
                        set.add(a[y]);
                        set.add(a[z]);
                        Collections.sort(set);
                        if (result.isEmpty()) result.add(set);
                        else {
                            boolean duplicate = false;
                            for (List<Integer> res : result) {
                                if (set.equals(res)) {
                                    duplicate = true;
                                    break;
                                }
                            }
                            if (!duplicate) result.add(set);
                        }
                    }
                }
            }
        }
        return result;
    }
}

File: src/com/leetcode/array/M15_Three_Sum.cc
#include "../common.h"
#include <algorithm>
using namespace std;

/**
 * Created by Michael on 2019/9/4.
 * 
 * Given an array nums of n integers, 
 * are there elements a, b, c in nums such that a + b + c = 0? 
 * Find all unique triplets in the array which gives the sum of zero.
 * 
 * Note:
 * The solution set must not contain duplicate triplets.
 * 
 * Example:
 * Given array nums = [-1, 0, 1, 2, -1, -4],
 * A solution set is:
 * [
 *      [-1, 0, 1],
 *      [-1, -1, 2]
 * ]
 * 
 */

/**
 * 首先，沿用 2sum 的思路，用 target - 元素值，得到 2sum 的新 target
 * 然后，利用已排序数组的性质，用双指针确定剩下两个元素的组合
 *
 * 问题一：为什么双指针选择首尾对向扫描 [→ ←]，而不是起始位置同向扫描 [→ →] ？
 * 思考过程：双指针该按什么顺序扫描，要看哪种扫描方式能够带来<可划分的确定趋势>
 * 对于已排序数组，左右指针同向扫描，没有什么区分度，右移左指针和右移右指针并不会带来截然不同的后果，因此我们无法形成一个移动指针的逻辑，也就不能这么做
 * 而如果是左右指针分别从左右边界扫描，那么右移左指针一定会使 2sum 变大，而左移右指针一定会让 2sum 变小，因此我们就得到了一个移动指针的逻辑，可以实现为代码
 * 
 * 问题二：如何避免重复
 * 思考过程：重复 triplets 的产生，唯一的可能来自于重复的元素，因此我只要跳过这些重复的元素，就一定可以避免重复的三元组。
 * 对于已排序数组，重复元素一定是相邻的，因此可以通过指针比较相邻元素来快速飘过重复元素。
 * 
 */

// Time - O(N^2)
// 外循环确定三元组的第一个元素
// 内循环用双指针确定剩下两个元素
vector<vector<int>> threeSum(vector<int>& nums) {
    vector<vector<int>> results;
    if (nums.size() < 3) return results;
    std::sort(nums.begin(), nums.end());                 // 先排序，后利用已排序数组的特性
    for (int i = 0; i < nums.size() - 2; i++) {          // 给后面两个指针留两个空位
        if (i > 0 && nums[i] == nums[i - 1]) continue;   // 去重：首先要跳过重复元素
        int target = 0 - nums[i];
        int left = i + 1;
        int right = nums.size() - 1;
        while (left < right) {
            if      (nums[left] + nums[right] < target) left++;     // 变得更大
            else if (nums[left] + nums[right] > target) right--;    // 变得更小
            else {
                results.push_back({nums[i], nums[left], nums[right]});
                left++;     // 继续寻找
                right--;    // 继续寻找
                while (left < right && nums[left] == nums[left - 1]) left++;        // 去重
                while (left < right && nums[right] == nums[right + 1]) right--;     // 去重
            }
        }
    }
    return results;
}

int main() {
    vector<int> v = {-1, 0, 1, 2, -1, -4};
    vector<vector<int>> r = threeSum(v);
    print2D(r);
}